ui/src/testing/react-test-helpers.ts

```diff
@@ -9,7 +9,8 @@ import * as fp from 'lodash/fp';
 // And fails with:
 //   ReferenceError: Zone is not defined
 export async function waitOneTickAndUpdate(wrapper: ReactWrapper) {
-  const waitImmediate = () => new Promise<void>(resolve => setImmediate(resolve));
+  // Make sure to use setTimeout. Combining setTimeout (used in UI) and setImmediate can result in a non-deterministic order of events
```

Comment:
Which specific callbacks are executing in a non-deterministic order in this test case?

Reply:
Specifically the [subscribable](https://github.com/all-of-us/workbench/blob/b31d28d9f8dfd531741507952e63edf6027ea4a3/ui/src/app/utils/subscribable.ts#L47) `setTimeout`:

Running this snippet in node repeatedly will demonstrate the issue
```
async function immediatePromise(v) {
  return new Promise(resolve => setImmediate(resolve));
}
async function timeoutPromise() {
  return new Promise(resolve => setTimeout(resolve, 0));
}
function immediate(v) {
  setImmediate(() => console.log('Immediate: ', v))
}
function timeout(v) {
  setTimeout(() => console.log('Timeout: ', v), 0)
}
async function main () {
  console.log('Start Run'.padStart(25, '-').padEnd(50, '-'))
  for (let i = 0; i < 10; i++) {
    // You can play around with different combinations of timeout/immediate to see the different behaviors
    // In general immediate is not deterministic when combined with timeout
    timeout(i)
    immediate(i)
    // await timeoutPromise() 
    // await immediatePromise()
    // console.log('After Immediate: ', i)
  }
}
main();
```

Reply:
Interesting, thanks for digging into this - analysis looks right to me, though I'm very confused as to how we haven't seen this issue before now.

Reply:
I think this is only really an issue when `setTimeout` has a delay of `0`. There aren't too many places in the code where this occurs. Cohort builder seems to do this a bit more than our other components. Most of our timeouts have a delay associated with them.